### PR TITLE
DAOS-10365 dfuse: Allow dfuse caching of metadata.

### PR DESCRIPTION
**Pull request by ashleypittman**
Stash the stat information for inodes and if the kernel re-requests the data within
the validity timeout then reply with the existing data and a timeout value to match
the original expiry date.
This is mostly for when writeback caching is enabled where the kernel will make
requests for metadata updates despite being given permission to cache the data.
Only use this cache when the file hasn't been written to since last update, and
disable when interception library is in use.

### CODE REVIEW

**Comment by ashleypittman** [comment]
This is failing in CI however I've verified what should be the same I/O pattern locally and it passes 20/20 times.  Still investigating.

**Review by daltonbohning — APPROVED** [review]

**Comment by ashleypittman** [comment]
This PR is also causing intermittent NLT failures with test_truncate_with_caching

File: src/client/dfuse/dfuse.h [hunk 1]
```diff
@@ -443,30 +453,27 @@ struct fuse_lowlevel_ops dfuse_ops;
...
-		__rc = fuse_reply_create(req, &entry, fi);		\
-		if (__rc != 0)						\
-			DFUSE_TRA_ERROR(desc,				\
-					"fuse_reply_create returned %d:%s",\
-					__rc, strerror(-__rc));		\
+#define DFUSE_REPLY_CREATE(desc, req, entry, fi)                                                   \
+	do {                                                                                       \
+		int __rc;                                                                          \
+		DFUSE_TRA_DEBUG(desc, "Returning create");                                         \
+		dfuse_cache_set_time(desc);                                                        \
```
Comment (ashleypittman) [inline src/client/dfuse/dfuse.h:460]: I think we should probably remove this.

**Review by jolivier23 — DISMISSED** [review]

**Review by daltonbohning — DISMISSED** [review]

**Review by mchaarawi — CHANGES_REQUESTED** [review]
i still have not had a chance to retry after the latest update, so hold on before landing

**Comment by ashleypittman** [comment]
This has passed a key test in CI, please can I request re-review.

**Review by mchaarawi — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**